### PR TITLE
OADP-3038: correct error when region cannot be auto discovered by velero

### PR DESCRIPTION
**Pull request by kaovilai**
The error message is now
"region for AWS backupstoragelocation not automatically discoverable. Please set the region in the backupstoragelocation config"
when it is not possible to automatically discover region for an aws bucket.

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

plugin region fix: https://github.com/openshift/openshift-velero-plugin/pull/223

### CODE REVIEW

**Comment by mateusoliveira43** [comment]
this PR fixes this flake https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_oadp-operator/1205/pull-ci-openshift-oadp-operator-master-4.11-operator-e2e-aws/1719799906991869952/artifacts/operator-e2e-aws/e2e/build-log.txt ? Have seen this 3 times I think in the last 2 weeks

**Comment by kaovilai** [comment]
> this PR fixes this flake https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_oadp-operator/1205/pull-ci-openshift-oadp-operator-master-4.11-operator-e2e-aws/1719799906991869952/artifacts/operator-e2e-aws/e2e/build-log.txt ? Have seen this 3 times I think in the last 2 weeks

Nope. Not related to any flake.

**Comment by kaovilai** [comment]
Note there are only string changes in this PR.

/unhold

**Review by mateusoliveira43 — APPROVED** [review]

**Comment by weshayutin** [comment]
/hold 
this bug is aligned to 1.3.2

**Comment by kaovilai** [comment]
/unhold
master is no longer used for 1.3
/cherry-pick oadp-1.3
will have to hold for 1.3.2 there.

**Comment by kaovilai** [comment]
```

  [FAILED] Expected
      <[]string | len:2, cap:2>: [
          "time=\"2023-11-27T23:44:03Z\" level=warning msg=\"VolumeSnapshot has a temporary error Failed to create snapshot: failed to remove VolumeSnapshotBeingCreated annotation on the content snapcontent-139d69df-ba6a-4da6-8b6e-c2e945e14d3a: \\\"snapshot controller failed to update snapcontent-139d69df-ba6a-4da6-8b6e-c2e945e14d3a on API server: Operation cannot be fulfilled on volumesnapshotcontents.snapshot.storage.k8s.io \\\\\\\"snapcontent-139d69df-ba6a-4da6-8b6e-c2e945e14d3a\\\\\\\": the object has been modified; please apply your changes to the latest version and try again\\\". Snapshot controller will retry later.\" backup=openshift-adp/mysql-twovol-csi-e2e-ad27f8a5-8d7e-11ee-b39b-0a580a81b464 cmd=/plugins/velero-plugin-for-csi logSource=\"/opt/app-root/src/velero-plugin-for-csi/internal/backup/volumesnapshot_action.go:250\" pluginName=velero-plugin-for-csi",
          "time=\"2023-11-27T23:44:03Z\" level=error msg=0 backup=openshift-adp/mysql-twovol-csi-e2e-ad27f8a5-8d7e-11ee-b39b-0a580a81b464 logSource=\"pkg/controller/backup_controller.go:729\"",
      ]
  to equal
      <[]string | len:0, cap:0>: []
  In [It] at: /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_suite_test.go:158 @ 11/27/23 23:51:32.565
  < Exit [It] MySQL application two Vol CSI - /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_suite_test.go:287 @ 11/27/23 23:51:32.565 (8m52.645s)
```
/retest

**Comment by weshayutin** [comment]
@kaovilai got conflicts :)

**Comment by kaovilai** [comment]
The changes here was blocked by https://github.com/openshift/openshift-velero-plugin/pull/223 which has since been merged.

**Comment by kaovilai** [comment]
after rebasing this closed (nothing to compare).. checking to see if further changes are required.

File: pkg/storage/aws/s3_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func TestBucketRegionIsDiscoverable(t *testing.T) {
+	type args struct {
+		bucket string
+	}
+	tests := []struct {
+		name string
+		args args
+		want bool
```
Comment (mateusoliveira43) [inline pkg/storage/aws/s3_test.go:64]: want seems a hard to understand name. Want what? Maybe wantError would be better (also, adding the check if error message is the expected, I think would be very valuable)

File: pkg/storage/aws/s3_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
+package aws
+
+import (
+	"testing"
+)
+
+// from https://github.com/openshift/openshift-velero-plugin/pull/223/files#diff-4f17f1708744bd4d8cb7a4232212efa0e3bfde2b9c7b12e3a23dcc913b9fc2ec
+func TestGetBucketRegion(t *testing.T) {
+	type args struct {
```
Comment (mateusoliveira43) [inline pkg/storage/aws/s3_test.go:9]: Is this used (if yes, make it high level, so both tests use the same struct)

File: pkg/storage/aws/s3.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+		if region != "" {
+			return region, nil
+		}
+	}
+
+	return "", errors.New("unable to determine bucket's region")
+}
+
+// For tests to use when discoverable bucket is required.
```
Comment (mateusoliveira43) [inline pkg/storage/aws/s3.go:47]: if this is only for tests, I would define it inside test file

File: pkg/storage/aws/s3.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+func BucketRegionIsDiscoverable(bucket string) bool {
+	_, err := GetBucketRegion(bucket)
+	return err == nil
+}
+
+// GetBucketRegion returns the AWS region that a bucket is in, or an error
+// if the region cannot be determined.
+// copied from https://github.com/openshift/openshift-velero-plugin/pull/223/files#diff-da482ef606b3938b09ae46990a60eb0ad49ebfb4885eb1af327d90f215bf58b1
```
Comment (mateusoliveira43) [inline pkg/storage/aws/s3.go:21]: here we can't import plugin because of cyclic import?

Comment (mateusoliveira43) [reply pkg/storage/aws/s3.go:21]: but plugin imports oadp, so we could tell it to import from here the function, and avoid duplicating code, no?

File: pkg/storage/aws/s3.go (hunk 3 above)
Comment (kaovilai) [reply pkg/storage/aws/s3.go:47]: Only work within a single package test, wanted this to be usable across different packages

File: pkg/storage/aws/s3.go (hunk 4 above)
Comment (kaovilai) [reply pkg/storage/aws/s3.go:21]: Yeah importing plugin go package would bump k8s

Comment (kaovilai) [reply pkg/storage/aws/s3.go:21]: Which breaks @shubham-pampattiwar knows the pain

File: pkg/storage/aws/s3_test.go (hunk 1 above)
Comment (kaovilai) [reply pkg/storage/aws/s3_test.go:64]: If you look at the function name tested it's self explanatory. TestBucketRegionIsDiscoverable
Want true or false.

File: pkg/storage/aws/s3.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+import (
+	"context"
+
+	"errors"
+
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	// "github.com/pkg/errors" don't need stack trace, not inside velero-plugin
```
Comment (kaovilai) [inline pkg/storage/aws/s3.go:11]: Delete comment.
```suggestion
```

File: controllers/bsl.go [hunk 6]
```diff
@@ -285,10 +286,14 @@ func (r *DPAReconciler) validateAWSBackupStorageLocation(bslSpec velerov1.Backup
...
 	}
 
-	// BSL region is required when s3ForcePathStyle is true OR BackupImages is true
-	// TODO: Update logic when (region auto-discovery) https://github.com/openshift/openshift-velero-plugin/pull/223 is merged
-	if (bslSpec.Config == nil || len(bslSpec.Config[Region]) == 0) && (bslSpec.Config != nil && bslSpec.Config[S3ForcePathStyle] == "true" || dpa.BackupImages()) {
-		return fmt.Errorf("region for AWS backupstoragelocation cannot be empty when s3ForcePathStyle is true or when backing up images")
+	// BSL region is required when
+	// - s3ForcePathStyle is true, because some velero processes requires region to be set and is not auto-discoverable when s3ForcePathStyle is true
+	//   imagestream backup in openshift-velero-plugin now uses the same method to discover region as the rest of the velero codebase
+	//   after https://github.com/openshift/openshift-velero-plugin/pull/223.
```
Comment (shubham-pampattiwar) [inline controllers/bsl.go:292]: NIT: Lets remove this comment as the PR is merged ?

**Review by shubham-pampattiwar — APPROVED** [review]

File: controllers/bsl.go (hunk 6 above)
Comment (kaovilai) [reply controllers/bsl.go:292]: ack

**Review by shubham-pampattiwar — APPROVED** [review]

**Review by mateusoliveira43 — APPROVED** [review]

**Pull request merged**